### PR TITLE
[Distributed] ResultHandler.onReturn must be ad-hoc because SerializationRequirement

### PR DESCRIPTION
**Pull request by ktoso**
Work in progress on last API that needs special treatment because of SerializationRequirement not being expressible as constraint.

UPDATE: Done, thank you @hamishknight for all the help here. Was quite a few tricky things here.
This is now working and just needs cleanup of some tests -- I'll do that in the morning. Thanks again 🙇 

Resolves rdar://90756969

---

Status is we're almost done with synthesis but facing issues with generics of the local function.

To reproduce, run `Distributed/Runtime/distributed_actor_func_calls_remoteCall_echo.swift`.

We currently crash on very weird nullptr on verifying the generics in TypeChecker::checkReferencedGenericParams.

The inner func looks good:

```
(func_decl implicit "doInvokeOnReturn(returnType:)" <R> interface type='<τ_0_0 where τ_0_0 : Decodable, τ_0_0 : Encodable> (returnType: R.Type) async throws -> ()'
  (parameter_list
    (parameter "returnType" apiName=returnType type='τ_0_0.Type' interface type='R.Type')))
```

but:

```
frame #1: 0x000000010277a814 swift-frontend`swift::TypeChecker::checkReferencedGenericParams(dc=0x0000000151011038) at TypeCheckGeneric.cpp:493:20
   490 	  // among referencedGenericParams.
   491 	  for (auto *genParam : genericSig.getGenericParams()) {
   492 	    auto *paramDecl = genParam->getDecl();
-> 493 	    if (paramDecl->getDepth() != fnGenericParamsDepth)
```

crashes as `paramDecl` is null here, very confusing 🤔 

I also tried to see if we could synthesize the type-checked AST here; which does not use `_openExistential` at all, but directly does `open_existential_expr` but that synthesis looked pretty hard... So trying with the parse equivalent for now.

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please test

**Comment by ktoso** [comment]
@swift-ci please test

**Comment by ktoso** [comment]
Resolves rdar://90756969

File: all.txt [hunk 1]
```diff
@@ -0,0 +1,242 @@
+-> % pbpaste > all.txt
```
Comment (xedin) [inline all.txt:1]: Looks like this has been added by mistake.

File: lib/AST/ASTDumper.cpp [hunk 2]
```diff
@@ -2625,11 +2625,11 @@ class PrintExpr : public ExprVisitor<PrintExpr> {
     printCommon(E, name) << ' ';
     if (auto checkedCast = dyn_cast<CheckedCastExpr>(E))
       OS << getCheckedCastKindName(checkedCast->getCastKind()) << ' ';
-    OS << "writtenType='";
-    if (GetTypeOfTypeRepr)
-      GetTypeOfTypeRepr(E->getCastTypeRepr()).print(OS);
-    else
-      E->getCastType().print(OS);
+        OS << "writtenType='";
```
Comment (xedin) [inline lib/AST/ASTDumper.cpp:2628]: This is incorrect code alignment.

Comment (ktoso) [reply lib/AST/ASTDumper.cpp:2628]: Whoops thanks. Should have run formatted again after the last commit -- will fix

File: lib/Sema/DerivedConformanceDistributedActor.cpp [hunk 3]
```diff
@@ -94,6 +100,321 @@ static FuncDecl *deriveDistributedActor_resolve(DerivedConformance &derived) {
...
+////    resultBuffer: UnsafeRawPointer,
+////    metatype _metatype: Any.Type
+////  ) async throws
+/// \endverbatim
+static FuncDecl *deriveDistributedActorSystem_invokeHandlerOnReturn(
+    DerivedConformance &derived) {
+  auto system = derived.Nominal;
+  auto &C = system->getASTContext();
+
+  auto mkParam = [&](Identifier argName, Identifier paramName,
```
Comment (xedin) [inline lib/Sema/DerivedConformanceDistributedActor.cpp:367]: Looks like this is where newly added `ParamDecl::createImplicit` could come handy.

File: include/swift/AST/Decl.h [hunk 4]
```diff
@@ -3500,12 +3500,17 @@ class NominalTypeDecl : public GenericTypeDecl, public IterableDeclContext {
...
   ConstructorDecl* getDistributedRemoteCallTargetInitFunction() const;
 
-  /// Find the 'RemoteCallArgument(label:name:value:)' initializer function
+  /// Find the 'RemoteCallArgument(label:name:value:)' initializer function.
   ConstructorDecl* getDistributedRemoteCallArgumentInitFunction() const;
 
+  /// Find the
+  /// 'DistributedActorSystem.invokeHandlerOnReturn(handler:value:metatype:)
+  /// function.
+  FuncDecl *getDistributedActorSystemInvokeHandlerOnReturnFunction() const;
```
Comment (xedin) [inline include/swift/AST/Decl.h:3512]: Is this a leftover code that has to be deleted? I don't see any usages of it and request associated with this method returns `nullptr` always with a `// FIXME`.

Comment (ktoso) [reply include/swift/AST/Decl.h:3512]: Ouch thanks for noticing; yes this is leftover indeed.

**Comment by ktoso** [comment]
Thanks for review! doing a mini cleanup PR to address those

**Pull request merged**